component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java

```diff
@@ -0,0 +1,281 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.am.analytics.publisher.retriever;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.moesif.api.MoesifAPIClient;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.am.analytics.publisher.exception.APICallException;
+import org.wso2.am.analytics.publisher.reporter.moesif.util.MoesifKeyEntry;
+import org.wso2.am.analytics.publisher.reporter.moesif.util.MoesifMicroserviceConstants;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.net.ssl.HttpsURLConnection;
+
+/**
+ * Responsible for calling the Moesif microservice and refresh/init entire internal map(organization, moesif key) or
+ * retrieving single moesif key.
+ * Also ease Moesif SDK client initiation by maintaining a map(moesif key, moesif sdk client {@link MoesifAPIClient}).
+ */
+public class MoesifKeyRetriever {
+    private static final Logger log = LoggerFactory.getLogger(MoesifKeyRetriever.class);
+    private static MoesifKeyRetriever moesifKeyRetriever;
+    private ConcurrentHashMap<String, String> orgIDMoesifKeyMap;
+
+    private ConcurrentHashMap<String, MoesifAPIClient> moesifKeyClientMap;
+    private String gaAuthUsername;
+    private String gaAuthPwd;
+
+    private MoesifKeyRetriever(String authUsername, String authPwd) {
+
+        this.gaAuthUsername = authUsername;
+        this.gaAuthPwd = authPwd;
+        orgIDMoesifKeyMap = new ConcurrentHashMap();
+        moesifKeyClientMap = new ConcurrentHashMap();
+    }
+
+    public static synchronized MoesifKeyRetriever getInstance(String authUsername, String authPwd) {
+        if (moesifKeyRetriever == null) {
+            return new MoesifKeyRetriever(authUsername, authPwd);
+        }
+        return moesifKeyRetriever;
+    }
+
+    /**
+     * Will initialize the empty orgID-MoesifKey map.
+     * Will refresh/refill the  orgID-MoesifKey map.
+     */
+    public void initOrRefreshOrgIDMoesifKeyMap() {
+        int attempts = MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS;
+        try {
+            callListResource();
+        } catch (IOException | APICallException ex) {
+            // TODO: Separate retry logic to a separate class.
+            log.error("First attempt failed,retrying.", ex);
```

Comment:
Shall we add more context in to the error message

Reply:
Within the method we refresh/init the internal map
So included that in the message.